### ISSUE TITLE
Interactions with sokol_app windows have low refresh rate in windows-10

### ISSUE THREAD

**Issue by kaervin** [floooh/sokol#531]
Hello,
Window-interactions, such as dragging the sokol-app-window, cause sokol to swap with lower framerate.
Measuring the framerate with sokol_time, it falls from around 144hz to unstable 50hz.

Additionally, the movement of the window itself becomes very unsmooth. 
The window position seems to only get updated at around 50hz, and not the full 144hz of my monitor. 
All other windows behave normally when dragging them around, and I have only noticed this issue with sokol_app windows.

To test this I used the sokol_app "clear sample", and instrumented it with measuring frametime every so often with sokol_time.

Changing the swap-interval doesn't change this behaviour.
I have one 1440p 144hz Monitor connected.
If you need any more info, please tell me what I can provide.

Thank you.

**Comment by floooh** [comment]
Yeah, I guess this is on NVIDIA? 

As far as I've seen this happens on NVIDIA cards with recent graphics drivers. Started happening around December 2020, the previous drivers were fine (pre v460). I'm currently sticking with v457 since the 46x drivers also make trouble with various PC games.

The problematic code is here, this continues rendering during move and resize operations:

https://github.com/floooh/sokol/blob/8f39840687d2675c199c6b34d6b3c64ba4a8786b/sokol_app.h#L6202-L6215

Removing this code fixes the problem here, but this of course means, the application is frozen until the move- and resize-operation ends.

I haven't found a good fix for this so far unfortunately, I was hoping that the issue will be fixed by NVIDIA (because the 46x drivers have generally been such a mess, but I guess after more than half a year, chances for this are slim).

**Comment by kaervin** [comment]
Yes, this is with a NVIDIA card.
Wouldn't think the graphics card drivers could have such an effect.
Trying it out with a game I can notice a similar behaviour, so it is not just sokol.

Some applications such as  Chrome or Blender don't have this problem though. 
Maybe they are doing something different, which prevents this problem from occurring?

**Comment by floooh** [comment]
One thing I haven't tried out so far is doing an explicit D3D11DeviceContext::Flush() in the move-resize-loop, this might help with memory spikes, and maybe it also has an effect on this specific problem (complete guess though).

**Comment by kaervin** [comment]
Excuse my lack of understanding, but I don't really get why creating a timer is needed in WM_ENTERSIZEMOVE. 
Why doesn't the loop in _sapp_win32_run just continue to call _sapp_frame(); and present the framebuffer?

After expirementing a bit, removing the timer, and adding the following case to _sapp_win32_wndproc seems to work much better than the timer-method.
```            
case WM_MOVE:
                _sapp_frame();
                #if defined(SOKOL_D3D11)
                    _sapp_d3d11_present();
                #endif
                #if defined(SOKOL_GLCORE33)
                    _sapp_wgl_swap_buffers();
                #endif
                break;
 ```
 It still doesn't run at full 144hz, but it is at least fluid enough with around 120hz.
 The problem here is that it runs the frame only when the window is actually moved.
 When I press the titlebar, without moving the window, sapp_frame() is not called.

My guess is, that the WM_TIMER event itself is just issued too infrequently to couple it to the frame-logic.
This brings me back to the question from above, which is why can't just the usual frame-loop continue to do its thing when WM_ENTERSIZEMOVE starts?
Is there any alternative sokol could use?

**Comment by kaervin** [comment]
After reading a bit about it, I now understand why the timer is needed.
It seems that the WM_SYSCOMMAND with SC_MOVE processed by DefWindowProcW just blocks the thread and continues to pump and process messages until the titlebar is released.
So the only way to get messages into the system which can call the frame-logic seem to be timers.
This is unfortunate, since the timer-events can be rather infrequent.

Processing the frame on WM_MOVE additionally to the timer seems to work ok enough.
Dragging around the sokol-window feels a lot smoother this way.
But the framerate is still limited when holding the titlebar and NOT moving the window. (This case is probably negligible).

Apart from this, it seems that the WM_ENTERSIZEMOVE event seems to have some lag too.
Here is a possible fix for this:
https://gamedev.net/forums/topic/672094-keeping-things-moving-during-win32-moveresize-events/5254386/

**Comment by floooh** [comment]
That's a good find, thanks!

**Comment by kaervin** [comment]
After having experimented a bit, I think having the app loop in a seperate thread from the window-handling is the way to go.
Interacting with the window has no implications on the app-loop this way.

Additionally adding a mutex, which the app-loop takes until it has called swap-buffers, and which WM_SIZING and WM_PAINT wait on, results in really stable resizing, where the window contents are never invalidated in any way.
The window will always only draw the framebuffer and no additional pixels.
The downside is that the resizing-rate of the window is then coupled to the frame-rate of the application.
I think this is better than having artifacts. Especially for gui-applications or multi-window stuff.

Additionally, this technique has the advantage that gpu-copy is actually reduced a bit when resizing.

**Comment by floooh** [comment]
Ok, so I have just committed a "fix/hack/workaround" which improves the situation here quite a bit (Win10, RTX 2070, latest drivers):

Basically, during window movement and resize (when WM_TIMER is active) the presentation happens with DXGI_PRESENT_DO_NOT_WAIT. Window movement is now always smooth here, and resizing "mostly" smooth (apart from some one-time hickups):

https://github.com/floooh/sokol/commit/dd9a1eac4efaff15e94c17896e32e3b803e36efc

...and I also implemented the 0.5sec freeze workaround you found:

https://github.com/floooh/sokol/commit/c9a36146320aa49ac79e44467d98d9eca31243f1

I thing both combined are a good-enough fix to close the ticket.

**Comment by kaervin** [comment]
Nice, thank you.
